File: ObjCVerify/ObjCVerify/Examples/Controller/离线方法调用/WYOffLineMethodController.m
//
//  WYOffLineMethodController.m
//  ObjCVerify
//
//  Created by guanren on 2026/1/13.
//

#import "WYOffLineMethodController.h"
#import <objc/message.h>
#import <WYBasisKitObjC/WYBasisKitObjC.h>

@interface WYOffLineMethodController ()

@end

@implementation WYOffLineMethodController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    Class class = NSClassFromString(@"WYGenericTypeController");
    if ([class isSubclassOfClass:[UIViewController class]]) {
        UIViewController *moduleClass = [[class alloc] init];
        NSString *method = @"testMethodWithData:data2:";
        SEL selector = NSSelectorFromString(method);
        
        if ([moduleClass respondsToSelector:selector]) {
            // 声明函数指针类型与所需参数
            id (*typed_msgSend)(id, SEL, NSString *, NSInteger) = (id (*)(id, SEL, NSString *, NSInteger))objc_msgSend;
            // 直接调用并传参
            typed_msgSend(moduleClass, selector, @"context", 99999);
        }
    }
}

/*
 #pragma mark - Navigation
 
 // In a storyboard-based application, you will often want to do a little preparation before navigation
 - (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
 // Get the new view controller using [segue destinationViewController].
 // Pass the selected object to the new view controller.
 }
 */

@end

File: SwiftVerify/SwiftVerify/Examples/Controller/离线方法调用/WYOffLineMethodController.swift
//
//  WYOffLineMethodController.swift
//  WYBasisKitTest
//
//  Created by 官人 on 2024/9/4.
//  Copyright © 2024 官人. All rights reserved.
//

import UIKit

class WYOffLineMethodController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
        
        let className = String(describing: WYGenericTypeController.self)
        let namespace = Bundle.main.infoDictionary!["CFBundleExecutable"] as! String
        let fullClassName = namespace + "." + className

        if let classType = NSClassFromString(fullClassName) as? UIViewController.Type {
            let obj = classType.init()
            
            let selector = NSSelectorFromString("testMothodWithData:data2:")
            
            if obj.responds(to: selector) {
                
                typealias Func = @convention(c) (AnyObject, Selector, NSString, Int) -> Unmanaged<AnyObject>?
                
                let imp = obj.method(for: selector)
                let function = unsafeBitCast(imp, to: Func.self)
                
                _ = function(obj, selector, "context", 99999)
            }
        }
    }
    

    /*
    // MARK: - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        // Get the new view controller using segue.destination.
        // Pass the selected object to the new view controller.
    }
    */

}
